# Loan Repayment Behaviour Analysis

OBJECTIVES:

    1. Identify borrower's characteristics leading to loan default.
    2. Highlight segments with high risk.
    3. Insights that could help lenders reduce risks.

In [1]:
import pandas as pd

In [2]:
#pd.set_option('display.max_columns', None)
#pd.set_option('display.width', None)
#pd.set_option('display.max_rows', None)

This lets us view the whole output without any truncation

In [3]:
df = pd.read_csv("C:/Loan_Repayment_Analysis/Data/accepted_2007_to_2018Q4.csv", nrows=50000, low_memory=False)

low_memory=False, makes Panda read the file in one pass before finalizing the data types thereby using more RAM
Else it would have read the file in chunks

In [4]:
df.shape

(50000, 151)

In [5]:
df.dtypes

id                         int64
member_id                float64
loan_amnt                float64
funded_amnt              float64
funded_amnt_inv          float64
                          ...   
settlement_status         object
settlement_date           object
settlement_amount        float64
settlement_percentage    float64
settlement_term          float64
Length: 151, dtype: object

In [6]:
df.columns.to_list()

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 '

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Columns: 151 entries, id to settlement_term
dtypes: float64(114), int64(1), object(36)
memory usage: 57.6+ MB


large datasets often contain incorrect datatypes

In [8]:
#Check for variables that might showcase the outcome
df.head()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,68407277,NaN,3600.0,3600.0,3600.0,36 months,13.99,123.03,C,C4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,68355089,NaN,24700.0,24700.0,24700.0,36 months,11.99,820.28,C,C1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,68341763,NaN,20000.0,20000.0,20000.0,60 months,10.78,432.66,B,B4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
3,66310712,NaN,35000.0,35000.0,35000.0,60 months,14.85,829.90,C,C5,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
4,68476807,NaN,10400.0,10400.0,10400.0,60 months,22.45,289.91,F,F1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [9]:
df.tail()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
49995,65753957,NaN,25000.0,25000.0,25000.0,36 months,6.24,763.28,A,A2,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
49996,65683967,NaN,13000.0,13000.0,13000.0,36 months,5.32,391.50,A,A1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
49997,65743096,NaN,7200.0,7200.0,7200.0,36 months,13.18,243.23,C,C3,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
49998,65252177,NaN,6000.0,6000.0,6000.0,36 months,8.18,188.52,B,B1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
49999,65603914,NaN,28000.0,28000.0,28000.0,36 months,5.32,843.22,A,A1,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [10]:
df['loan_status']

0         Fully Paid
1         Fully Paid
2         Fully Paid
3            Current
4         Fully Paid
            ...     
49995    Charged Off
49996     Fully Paid
49997     Fully Paid
49998     Fully Paid
49999     Fully Paid
Name: loan_status, Length: 50000, dtype: object

In [11]:
#Check for missing values
df.isnull()

,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
1,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
2,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
3,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
4,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
49995,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
49996,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
49997,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True
49998,False,True,False,False,False,False,False,False,False,False,...,True,True,False,False,True,True,True,True,True,True


In [12]:
missing_ratio = df.isnull().mean().sort_values(ascending=False)

In [13]:
missing_ratio

sec_app_open_acc           1.0
member_id                  1.0
revol_bal_joint            1.0
sec_app_fico_range_low     1.0
sec_app_fico_range_high    1.0
                          ... 
initial_list_status        0.0
mort_acc                   0.0
mo_sin_rcnt_tl             0.0
mo_sin_rcnt_rev_tl_op      0.0
id                         0.0
Length: 151, dtype: float64

In [14]:
#Summary of the dataset
summary = pd.DataFrame({
    "column" : df.columns,
    "dtype" : df.dtypes,
    "missing_%" : df.isnull().mean()*100
})
summary.sort_values("missing_%", ascending=False)

,column,dtype,missing_%
sec_app_open_acc,sec_app_open_acc,float64,100.0
member_id,member_id,float64,100.0
revol_bal_joint,revol_bal_joint,float64,100.0
sec_app_fico_range_low,sec_app_fico_range_low,float64,100.0
sec_app_fico_range_high,sec_app_fico_range_high,float64,100.0
...,...,...,...
initial_list_status,initial_list_status,object,0.0
mort_acc,mort_acc,float64,0.0
mo_sin_rcnt_tl,mo_sin_rcnt_tl,float64,0.0
mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_rev_tl_op,float64,0.0


Pre-loan variables:
    -income
    -employement length
    -home owernship
    -loan amount
    -interest rate
    -loan purpose

Post-loan variables:
    -payments
    -recoveries
    -collections
    -final loan status

In [15]:
df['loan_status'].value_counts()

loan_status
Fully Paid            34978
Charged Off            9027
Current                5610
Late (31-120 days)      246
In Grace Period         100
Late (16-30 days)        38
Default                   1
Name: count, dtype: int64

Censored Data --> means the final outcome has not occured yet.
Observations wherein the final outcome of the loans that are still "Current"
Gen. for such projects we keep only loans with known outcomes.

Rows to keep --> decide based on loan status
Col usable --> decide based on missing value ratio

We will remove columns that are 100% missing

# Data Cleaning
Now that we have explored the data we will be cleaning it 

In [16]:
empty_cols = missing_ratio[missing_ratio == 1].index

In [17]:
empty_cols

Index(['sec_app_open_acc', 'member_id', 'revol_bal_joint',
       'sec_app_fico_range_low', 'sec_app_fico_range_high',
       'sec_app_earliest_cr_line', 'sec_app_inq_last_6mths',
       'sec_app_mort_acc', 'sec_app_revol_util', 'sec_app_open_act_il',
       'sec_app_num_rev_accts', 'sec_app_chargeoff_within_12_mths',
       'sec_app_collections_12_mths_ex_med',
       'sec_app_mths_since_last_major_derog'],
      dtype='object')

In [18]:
df = df.drop(columns=empty_cols)

In [19]:
df.shape

(50000, 137)

Dropped 14 empty columns having missing values

In [20]:
#Removing Identifier columns
[col for col in df.columns if "id" in col]

['id']

In [21]:
df = df.drop(columns='id', errors="ignore")

We removed these identifier as these do not explain borrower behavior and cannot be used in analysis

In [22]:
df['int_rate'].dtype

dtype('float64')

In [23]:
df['int_rate'].head()

0    13.99
1    11.99
2    10.78
3    14.85
4    22.45
Name: int_rate, dtype: float64

In [24]:
df['term']

0         36 months
1         36 months
2         60 months
3         60 months
4         60 months
            ...    
49995     36 months
49996     36 months
49997     36 months
49998     36 months
49999     36 months
Name: term, Length: 50000, dtype: object

In [28]:
df['term'] = df['term'].str.extract(r"(\d+)").astype(int)

r here means raw string --> do not interpret backslashes. Changed the column into usable format

In [29]:
df['term']

0        36
1        36
2        60
3        60
4        60
         ..
49995    36
49996    36
49997    36
49998    36
49999    36
Name: term, Length: 50000, dtype: int32

In [30]:
df['issue_d']

0        Dec-2015
1        Dec-2015
2        Dec-2015
3        Dec-2015
4        Dec-2015
           ...   
49995    Nov-2015
49996    Nov-2015
49997    Nov-2015
49998    Nov-2015
49999    Nov-2015
Name: issue_d, Length: 50000, dtype: object

As it is a text format so we convert it into Datetime so later these could be used for time series analysis

In [31]:
df['issue_d'] = pd.to_datetime(df['issue_d'], format="%b-%Y")

The %b-%Y is a date representation format where %b is the abbreviated month name and %y is the two-digit year.

In [32]:
df['issue_d']

0       2015-12-01
1       2015-12-01
2       2015-12-01
3       2015-12-01
4       2015-12-01
           ...    
49995   2015-11-01
49996   2015-11-01
49997   2015-11-01
49998   2015-11-01
49999   2015-11-01
Name: issue_d, Length: 50000, dtype: datetime64[ns]

In [33]:
pd.set_option('display.max_columns', None)

In [34]:
pd.reset_option('display.max_columns')

In [35]:
df.head()

,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,3600.0,3600.0,3600.0,36,13.99,123.03,C,C4,leadman,10+ years,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,24700.0,24700.0,24700.0,36,11.99,820.28,C,C1,Engineer,10+ years,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,20000.0,20000.0,20000.0,60,10.78,432.66,B,B4,truck driver,10+ years,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
3,35000.0,35000.0,35000.0,60,14.85,829.90,C,C5,Information Systems Officer,10+ years,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
4,10400.0,10400.0,10400.0,60,22.45,289.91,F,F1,Contract Specialist,3 years,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [36]:
cols = ['last_pymnt_d', 'last_credit_pull_d']

In [37]:
df[cols] = df[cols].apply(pd.to_datetime, format="%b-%Y")

In [38]:
df[cols].dtypes

last_pymnt_d          datetime64[ns]
last_credit_pull_d    datetime64[ns]
dtype: object

In [39]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], format="%b-%Y")

In [40]:
#Checks for duplicate rows in the dataset, if it exists remove using drop_duplicates() function
df.duplicated().sum()

0

In [41]:
df['desc'].isnull().mean()

0.99994

In [42]:
df = df.drop(columns="desc")

Removed the column as it had almost 99.994% missing values. Might be an optional information that not all customers filled.

In [43]:
(df.isnull().mean() > 0.9).sum()

23

Won't drop all blindly as some columns are structurally sparse but meaningful so first inspect it.
As genrally columns with high missing values fall into three categories:
    1. Optional borrower input fields.   (Action --> Drop)
    2. Rare credit events.               (Action --> Often keep)  eg. mnths_since_last_record (These are missing because most borrowers never had those events)
    3. Servicing/operational fields      (Action --> Usually drop) eg. next_pymnt_d (Relates to loan servicing operation, not borrower risk)

In [44]:
high_missing_columns = df.columns[df.isnull().mean() > 0.9]

In [45]:
high_missing_columns

Index(['annual_inc_joint', 'dti_joint', 'verification_status_joint',
       'hardship_type', 'hardship_reason', 'hardship_status', 'deferral_term',
       'hardship_amount', 'hardship_start_date', 'hardship_end_date',
       'payment_plan_start_date', 'hardship_length', 'hardship_dpd',
       'hardship_loan_status', 'orig_projected_additional_accrued_interest',
       'hardship_payoff_balance_amount', 'hardship_last_payment_amount',
       'debt_settlement_flag_date', 'settlement_status', 'settlement_date',
       'settlement_amount', 'settlement_percentage', 'settlement_term'],
      dtype='object')

In [46]:
df['settlement_status'].describe()

count         1537
unique           3
top       COMPLETE
freq           654
Name: settlement_status, dtype: object

In [47]:
df = df.drop(columns=high_missing_columns)

On inspecting, the columns having high missing values were mostly those that were not essential for our this very particular objective so we dropped them all.

    1. Joint Borrower Variables : Used when two borrowers apply jointly for a loan which is rare specially in this dataset.
    2. Hardship Program Variables : Describe loan hardship programs which are rare so most of them have no records.
    3. Debt Settlement Variables : Appear when a defaulted loan enter settlement negotiations. These are post default events and aren't useful for EDA.

In [48]:
df.shape

(50000, 112)

#Now we will deal with Duplicate columns representing the same variables/informations.
& Columns that leak future information

# Four Categories of Columns in Credit Risk

1. Borrower Characteristics : These describe the person taking the loan. eg: annual_inc, emp_length, home_owernship, verification_status, addr_state
2. Loan Characteristics : These describe the loan contract. eg: loan_amnt, term, int_rate, grade, sub_grade, purpose, installment
3. Credit History Variables : These come from the borrowers credit bureau report. eg: dti, delinq_2yrs, open_acc, revol_bal, public_rec_bankruptcies. These measure debt levels, credit utilization, past delinquencies.
4. Post Loan/Repayment Variables : These describe what happened after loan started. eg: total_pymnt, recoveries, last_pymnt_d, total_rec_prncp, collection_recovery_fee.

This classification matters in making the decision of which variables to keep and which to remove. 

In [49]:
df.columns.to_list()

['loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'acc_now_delinq',
 'tot_coll_amt',
 'tot_cur_bal',
 

We remove post loan variables as the analysis becomes invalid because the model would cheat using future information. This is data leakage.
We also remove highly correlated variables. eg: loan_amnt, funded_amnt, funded_amnt_inv or say grade, sub-grade

For risk analysis at loan issuance only keep variables known at loan approval time.

In [50]:
[col for col in df.columns if 'pymnt' in col]

['pymnt_plan',
 'total_pymnt',
 'total_pymnt_inv',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d']

In [51]:
[col for col in df.columns if "rec" in col]

['mths_since_last_record',
 'pub_rec',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'mths_since_recent_bc',
 'mths_since_recent_bc_dlq',
 'mths_since_recent_inq',
 'mths_since_recent_revol_delinq',
 'pub_rec_bankruptcies']

In [52]:
#All these variables are known after loan begins repayment so they create data leakage
df = df.drop(columns=[
    "pymnt_plan",
    "total_pymnt",
    "total_pymnt_inv",
    "last_pymnt_d",
    "last_pymnt_amnt",
    "next_pymnt_d"
])

In [53]:
df.shape

(50000, 106)

In [54]:
#These occur after loan repayment begins
df = df.drop(columns=[
    "total_rec_prncp",
    "total_rec_int",
    "total_rec_late_fee",
    "recoveries",
    "collection_recovery_fee"
])

Variables the we kept from the list:
    pub_rec
    pub_rec_bankruptcies    
These are important risk indicators.
-number of derogatory public records
-number of bankruptcies


    mths_since_last_record (investigate but usually keep)
    mths_since_recent_bc
    mths_since_recent_bc_dlq
    mths_since_recent_inq
    mths_since_recent_revol_delinq
These measure recent credit behavior, which is useful for risk analysis.

In [55]:
#Removing Redundant variables
df = df.drop(columns=[
    'funded_amnt',
    'funded_amnt_inv',
    'sub_grade',
    'installment'
], errors="ignore")

In [56]:
df.shape

(50000, 97)

In [57]:
df['loan_status'].value_counts()

loan_status
Fully Paid            34978
Charged Off            9027
Current                5610
Late (31-120 days)      246
In Grace Period         100
Late (16-30 days)        38
Default                   1
Name: count, dtype: int64

For risk analysis we only keep the completed loans, i.e. loans with status : Fully Paid and Charged Off. Remove everything else.
Default is just 1 loan so its negligible. 

In [58]:
df = df[df['loan_status'].isin(['Fully Paid', 'Charged Off'])]

In [59]:
df.shape

(44005, 97)

-Converted the dataset from multi class status to binary status.  
-Keeping the ongoing loans would have made the calculation biased.

In [60]:
df['loan_status'].value_counts()

loan_status
Fully Paid     34978
Charged Off     9027
Name: count, dtype: int64

In [62]:
#Creating Binary Target Variable
df["default"] = (df['loan_status'] == "Charged Off").astype(int)

In [65]:
df['default'].mean()

0.20513578002499716

About 20% of completed loans resulted in default

In [71]:
df.columns

Index(['loan_amnt', 'term', 'int_rate', 'grade', 'emp_title', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'url', 'purpose', 'title', 'zip_code', 'addr_state',
       'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low',
       'fico_range_high', 'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'last_credit_pull_d', 'last_fico_range_high',
       'last_fico_range_low', 'collections_12_mths_ex_med',
       'mths_since_last_major_derog', 'policy_code', 'application_type',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_

Many of these 97 columns are metadata, identifiers, post-loan servicing variables, extremely granular credit bureau variables. Even though earlier we had removed some but still the number of variables is high so we perform feature reduction and reduce the dataset to clean analytical feature set of around 20-30 variables

In [72]:
#Removing Non-Analytical Metadata
df = df.drop(columns=[
    'emp_title',
    'url',
    'title',
    'zip_code',
    'policy_code',
])

Didn't remove addr_state as of now because sometimes the geographical location is of importance because they affect the local economic conditions, unemployement levels, cost of living and so on. In some cases these are dropped when the number of categories are very large and it is harder to visualize.

In [73]:
#Removing Post Loan Variables
df = df.drop(columns=[
    'out_prncp',
    'out_prncp_inv',
    'last_credit_pull_d',
    'last_fico_range_high',
    'last_fico_range_low',
    'hardship_flag',
    'debt_settlement_flag'
])

In [74]:
#Simplify FICO Score instead of keeping two var, fico_range_high and fico_range_low we keep one average fico_score
df['fico_score'] = (df['fico_range_high'] + df['fico_range_low']) / 2


In [75]:
df = df.drop(columns=['fico_range_high' , 'fico_range_low'])

In [76]:
df.shape

(44005, 85)

We keep the core analytical variables, variables of characteristics: Loan characteristics, Borrower characteristics, Credit Profile, Credit score and Target while also kept some credit indicators which can be useful : inq_last_6mths, collections_12_mths_ex_med, acc_now_delinq, tot_cur_bal, mort_acc

As our goal for this project is Insight Generation, not building a predictive model. So we now remove further more variables even though some of them might depict real credit risk modeling.

In [78]:
df.columns

Index(['loan_amnt', 'term', 'int_rate', 'grade', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs',
       'earliest_cr_line', 'inq_last_6mths', 'mths_since_last_delinq',
       'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status',
       'collections_12_mths_ex_med', 'mths_since_last_major_derog',
       'application_type', 'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal',
       'open_acc_6m', 'open_act_il', 'open_il_12m', 'open_il_24m',
       'mths_since_rcnt_il', 'total_bal_il', 'il_util', 'open_rv_12m',
       'open_rv_24m', 'max_bal_bc', 'all_util', 'total_rev_hi_lim', 'inq_fi',
       'total_cu_tl', 'inq_last_12m', 'acc_open_past_24mths', 'avg_cur_bal',
       'bc_open_to_buy', 'bc_util', 'chargeoff_within_12_mths', 'delinq_amnt',
       'mo_sin_old_il_acct', 'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op',
    

In [79]:
df = df.drop(columns=[
    'earliest_cr_line',
    'mths_since_last_delinq',
    'mths_since_last_record', 
    'initial_list_status',
    'collections_12_mths_ex_med',
    'mths_since_last_major_derog',
    'application_type',
    'acc_now_delinq',
    'tot_coll_amt',
    'open_acc_6m',
    'open_act_il',
    'open_il_12m',
    'open_il_24m',
    'mths_since_rcnt_il',
    'total_bal_il',
    'il_util',
    'open_rv_12m',
    'open_rv_24m',
    'max_bal_bc',
    'all_util',
    'total_rev_hi_lim',
    'inq_fi',
    'total_cu_tl',
    'inq_last_12m',
    'acc_open_past_24mths',
    'avg_cur_bal',
    'bc_open_to_buy',
    'bc_util',
    'chargeoff_within_12_mths',
    'delinq_amnt',
    'mo_sin_old_il_acct',
    'mo_sin_old_rev_tl_op',
    'mo_sin_rcnt_rev_tl_op',
    'mo_sin_rcnt_tl',
    'mort_acc',
    'mths_since_recent_bc',
    'mths_since_recent_bc_dlq',
    'mths_since_recent_inq',
    'mths_since_recent_revol_delinq',
    'num_accts_ever_120_pd',
    'num_actv_bc_tl',
    'num_actv_rev_tl',
    'num_bc_sats',
    'num_bc_tl',
    'num_il_tl',
    'num_op_rev_tl',
    'num_rev_accts',
    'num_rev_tl_bal_gt_0',
    'num_sats',
    'num_tl_120dpd_2m',
    'num_tl_30dpd',
    'num_tl_90g_dpd_24m',
    'tot_hi_cred_lim',
    'total_bal_ex_mort',
    'total_bc_limit',
    'total_il_high_credit_limit',
    'disbursement_method'
])

In [80]:
df.shape

(44005, 28)

In [81]:
df.columns

Index(['loan_amnt', 'term', 'int_rate', 'grade', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'purpose', 'addr_state', 'dti', 'delinq_2yrs',
       'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util',
       'total_acc', 'tot_cur_bal', 'num_tl_op_past_12m', 'pct_tl_nvr_dlq',
       'percent_bc_gt_75', 'pub_rec_bankruptcies', 'tax_liens', 'default',
       'fico_score'],
      dtype='object')